src/hotspot/share/oops/compressedKlass.hpp

```diff
@@ -97,42 +98,98 @@ class Klass;
-  // Note: guaranteed to be aligned to KlassAlignmentInBytes
+  // Note: guaranteed to be aligned to 1<<shift (klass_alignment_in_bytes)
   static address _klass_range_start;
   static address _klass_range_end;
 
+  // Values for the lowest (inclusive) and highest (inclusive) narrow Klass ID, given the
+  // current Klass Range and encoding settings.
+  static narrowKlass _lowest_valid_narrow_klass_id;
+  static narrowKlass _highest_valid_narrow_klass_id;
+
+
   // Helper function for common cases.
   static char* reserve_address_space_X(uintptr_t from, uintptr_t to, size_t size, size_t alignment, bool aslr);
   static char* reserve_address_space_for_unscaled_encoding(size_t size, bool aslr);
   static char* reserve_address_space_for_zerobased_encoding(size_t size, bool aslr);
   static char* reserve_address_space_for_16bit_move(size_t size, bool aslr);
+  static void calc_lowest_highest_narrow_klass_id();
 
-  DEBUG_ONLY(static void assert_is_valid_encoding(address addr, size_t len, address base, int shift);)
+#ifdef ASSERT
+  static void sanity_check_after_initialization();
+#endif // ASSERT
 
-  static inline Klass* decode_not_null_without_asserts(narrowKlass v, address base, int shift);
-  static inline Klass* decode_not_null(narrowKlass v, address base, int shift);
+  template <typename T>
+  static inline void check_init(T var) {
+    assert(var != (T)-1, "Not yet initialized");
+  }
 
-  static inline narrowKlass encode_not_null(Klass* v, address base, int shift);
+  static inline Klass* decode_not_null_without_asserts(narrowKlass v, address base, int shift);
 
 public:
 
+  // Initialization sequence:
+  // 1) Parse arguments. The following arguments take a role:
+  //      - UseCompressedClassPointers
+  //      - UseCompactObjectHeaders
+  //      - Xshare on off dump
+  //      - CompressedClassSpaceSize
+  // 2) call pre_initialize(): depending on UseCompactObjectHeaders, defines the limits of narrow Klass pointer
+  //    geometry (how many bits, the max. possible shift)
+  // 3) .. from here on, narrow_klass_pointer_bits() and max_shift() can be used
+  // 4) call reserve_address_space_for_compressed_classes() either from CDS initialization or, if CDS is off,
+  //    from metaspace initialization. Reserves space for class space + CDS, attempts to reserve such that
+  //    we later can use a "good" encoding scheme. Reservation is highly CPU-specific.
+  // 5) Initialize the narrow Klass encoding scheme by determining encoding base and shift:
+  //   5a) if CDS=on: Calls initialize_for_given_encoding() with the reservation base from step (4) and the
+  //       CDS-intrinsic setting for shift; here, we don't have any freedom to deviate from the base.
+  //   5b) if CDS=off: Calls initialize() - here, we have more freedom and, if we want, can choose an encoding
+  //       base that differs from the reservation base from step (4). That allows us, e.g., to later use
+  //       zero-based encoding.
```

Comment:
Yes, I think so. I think the SAP Jit people investigated this when doing the PPC ports. You save at least two instructions, and possibly more, per decode op. You save code size too since you don't need to materialize the 64-bit base immediate. Especially on x64 this can mean easily 11 fewer bytes.